### ISSUE TITLE
Add additional configuration options when using NIO Transport Services

### ISSUE THREAD

**Issue by AgentFeeble** [hummingbird-project/hummingbird-core#62]
I need to add a local HTTP server to an iOS app I'm working on, and this library looks like such a great way of implementing the service I need. Being an iOS app, I'll be using NIO Transport Services, the library is just missing support for a few configuration options I need.

I'd be happy to submit a PR adding support for this if you're open to it, but I'm quite a bit of a newbie in regards to SwiftNIO (and this library too), so I wouldn't mind a few hints to see if I'm heading in the right direction here.

Specifically, I need the server to be able to broadcast a Bonjour service, and to be able to allow peer-to-peer connections.

After looking through the source code for NIO Transport Services, it looks like all you need to do to advertise a Bonjour service is to bind the `NIOTSListenerBootstrap` to an `NWEndpoint.service` instead of a host/port or unix domain socket. To do this, I would need to add an `nwEndpoint` case to the `HBBindAddress` enum. (but only when Network can be imported), and handle that additional case in `HBHTTPServer.start(responder:)`.

For including peer-to peer connections, all that needs to change is adding an `enablePeerToPeer` option in `HBApplication.Configuration` (only if Network can be imported), and setting that option for the server and child channels in `HBHTTPServer.createTSBootstrap(quiesce:, childChannelInitializer:)`.

This framework looks really great by the way, nice work!

**Comment by adam-fowler** [comment]
Adding the bind address shouldn't be an issue
Adding additional options is a breaking change (ie requires a major version update) unless you add another function.

**Comment by AgentFeeble** [comment]
Thanks for the response!

Would you mind elaborating on the additional options being a breaking change though? I'm probably missing something obvious, but I'm not seeing how adding an option would be a breaking change if we give it a default in the various constructors. I'm also not quite seeing where a function to add the additional option would sit to prevent the breaking change,.

**Comment by adam-fowler** [comment]
So if someone holds a reference to `HBConfiguration.init`, adding a parameter to the function will mean their code won't compile anymore. It is a fairly rare case. Many libraries don't consider optional parameters as a breaking change, even though they are. If this was going to be a case where the new parameter was for a feature that wasn't so niche, I might be happier to let that through.

In actual fact adding an extra field to the HBBindAddress enum is breaking as well. SemVer is so painful. I'm happy to let this one through though.

Are you sure the server channel as well as the child channels needs the peerToPeer flag. Because if it was only the child channels you could create your own `HBChannelInitializer`. Copy `HTTP1ChannelInitializer` and add the `channel.setOption` at the top of the `initialize` function. Then set `HBHTTPServer.httpChannelInitializer` to point to your version of the channel initialiser.

**Comment by AgentFeeble** [comment]
Ah ok I see, that makes sense. I hadn't considered the situation where you hold a reference to the init function, or that Swift would see the two versions of the function types as incompatible. I know what you mean about SemVer. Thanks for allowing this change through though, I appreciate it.

I know that both the client `NWConnection` as well as the `NWListener` need to be marked as including peer-to-peer, but I'm not actually sure if NIO requires it to be specified for both server & child, or if it's specified as a child option then automatically add it the option to the server. I'll test it out, because that would be a lot simpler not having to update the library.

If it turns out that both the server channel and child channels need the flag, another option instead of adding the `includePeerToPeer` flag to the `HBHTTPServer.Configuration` would be adding a similar callback function to the `HBChannelInitializer`. A callback function could allow the server channel to be configured arbitrarily, without adding extra arguments to the configuration, and it could be added as a backward compatible change if adding it as a separate property to `HBHTTPServer`. 

Another option would be to allow for setting a list of arbitrary `(ChannelOption, Value)` pairs in the configuration, which are all applied to the server channel during initialization.

**Comment by AgentFeeble** [comment]
Ok I've done some testing, and it turns out that the peerToPeer flag _must_ be set on the server channel. I think the best way to handle this in a backward compatible way would be to add a property to the `HBHTTPServer` that can be used customize the bootstrap object before binding to an address.

While testing this out, I also ran into an issue adding the `nwEndpoint` case to `HBBindAddress`. I would need to add the case conditionally, so that it's only available on iOS 12+, MacOS 10.14+, etc. But Swift doesn't support `@available` annotations for enum cases that have associated values, so there isn't a way to provide this case conditionally on platforms/OS versions that support it.

How would you feel about adding a `custom` case to the binding address? Some thing like this:

```swift
public enum HBBindAddress {
    // ...hostname and unixDomainSocket cases...

    case custom((HTTPServerBootstrap) -> EventLoopFuture<Void>)

    // ... convenience functions
}
```

It would mean having to expose the `HTTPServerBootstrap` publicly though (or finding some other type to fit the bootstrap object into). Instead of the `custom` case's associated value being a function, it could also be a protocol that has a single method with the same function signature.

I know that this changes the semantics of the `HBBindAddress`, it isn't really an "address" any more if it can also be a function that uses a custom function to perform the bind operation. But it does open the library up to easier customization without much effort.

Implementing this `custom` binding address case will also mitigate my need to add a property to the `HBHTTPServer` that can specify the peerToPeer option, since I can just modify those options directly in the function that performs the binding operation.

I know that this is probably quite a niche request that doesn't align with the majority of the use cases for this library, so I understand if you'd rather not introduce a change like this though. But if you wouldn't mind these changes, let me know, and I can work on a pull request for it.

**Comment by adam-fowler** [comment]
Yeah you can go ahead and add a custom binding address. 
```swift
case custom((HTTPServerBootstrap) -> EventLoopFuture<Channel>
```
I guess if you extend `HTTPServerBootstrap` to include `childChannelOption` and `serverChannelOption` you can set your peerToPeer option.

**Comment by AgentFeeble** [comment]
Awesome, I'll work on a PR adding support for this.Thanks for the amazing library 😊

**Comment by adam-fowler** [comment]
PR is merged and has been released in 1.4.0
